tools/autograd/gen_python_functions.py

```diff
@@ -926,92 +1188,298 @@ def get_python_signature(declaration, include_out):
-        else:
-            # NB: For more than 1 output args the type name is a TensorList
-            # and as such we don't (yet) need to consider the naming.
-            py_formal_args.append(typename + ' out=None')
-
-    # we could put this in the loop above but we want to ensure both type dispatched args
-    # and python binding arguments are after the out argument; this matches the case
-    # where there is a python binding argument dtype, which is necessary to match
-    # the function signatures between the out and non-out variant.
-    assert len(type_args) <= 1
-    for arg in type_args:
-        if positional:  # assume type_args should be kwarg_only.
-            py_formal_args.append('*')
-            positional = False
-        py_formal_args.append(get_py_formal_arg(arg))
-
-    if len(declaration['python_binding_arguments']) > 0:
-        for arg in declaration['python_binding_arguments']:
-            if arg.get('kwarg_only', False) and positional:
-                py_formal_args.append('*')
-                positional = False
-            py_formal_args.append(get_py_formal_arg(arg))
+    return python_binding_arguments
 
-    # Python function signature.
-    # This is the string that we give to FunctionParameter, which is
-    # then parsed into the actual structure which we do parsing
-    # with.
-    return PYTHON_FUNCTION_SIGNATURE.substitute(name=name, py_formal_args=py_formal_args)
+#
+# declaration derived props, utils, etc.
+# declarations are dicts loaded from Declarations.yaml,
+# passed to our codegen methods by callers in gen_autograd
+#
+
+def is_tensor_self(arg):
+    return arg['name'] == 'self' and arg['simple_type'] == 'Tensor'
+
+
+def is_tensor_options(arg):
+    return arg['simple_type'] == 'TensorOptions'
+
+
+def is_scatter(arg):
+    return arg.get('scatter_args') is not None
+
+def is_output(arg):
+    return arg.get('output', False)
+
+
+def has_outputs(declaration):
+    return any([is_output(arg) for arg in declaration['arguments']])
+
+
+def get_tensor_options(declaration):
+    args = [arg for arg in declaration['arguments'] if is_tensor_options(arg)]
+    if len(args) == 0:
+        return None
+    if len(args) != 1:
+        raise RuntimeError(
+            '{}: multiple tensor options arguments'.
+            format(declaration['name']))
+    return args[0]
+
+
+def has_tensor_options(declaration):
+    return get_tensor_options(declaration) is not None
+
+
+def is_tensor_method(declaration):
+    return 'Tensor' in declaration['method_of']
+
+
+def is_torch_function(declaration):
+    return 'namespace' in declaration['method_of']
+
+
+def is_nn_module_function(declaration):
+    return declaration.get('python_module') == 'nn'
+
+
+def function_namespace(declaration):
+    if has_tensor_options(declaration) or op_name(declaration).endswith('_like'):
+        return 'torch'
+    else:
+        return 'at'
```

Comment:
Sure, added a comment. Tried `torch` everywhere just to check the easy change but there are definitely lots of ops that are `at` and not `torch` - but maybe you mean there might no longer be any reason not to move them over... FWIW here's the first 20 before the compiler gives up

```
no member named '_adaptive_avg_pool2d' in namespace 'torch'; did you mean 'at::_adaptive_avg_pool2d'?
no member named '_addr' in namespace 'torch'; did you mean 'at::_addr'?
no member named '_addr_out' in namespace 'torch'; did you mean 'at::_addr_out'?
no member named '_addr_' in namespace 'torch'; did you mean 'at::_addr_'?
no member named '_baddbmm_mkl_' in namespace 'torch'; did you mean 'at::_baddbmm_mkl_'?
no member named '_batch_norm_impl_index' in namespace 'torch'; did you mean 'at::_batch_norm_impl_index'?
no member named '_cast_Byte' in namespace 'torch'; did you mean 'at::_cast_Byte'?
no member named '_cast_Char' in namespace 'torch'; did you mean 'at::_cast_Char'?
no member named '_cast_Double' in namespace 'torch'; did you mean 'at::_cast_Double'?
no member named '_cast_Float' in namespace 'torch'; did you mean 'at::_cast_Float'?
no member named '_cast_Half' in namespace 'torch'; did you mean 'at::_cast_Half'?
no member named '_cast_Int' in namespace 'torch'; did you mean 'at::_cast_Int'?
no member named '_cast_Long' in namespace 'torch'; did you mean 'at::_cast_Long'?
no member named '_cast_Short' in namespace 'torch'; did you mean 'at::_cast_Short'?
no member named '_cat' in namespace 'torch'; did you mean 'at::_cat'?
no member named '_cat_out' in namespace 'torch'; did you mean 'at::_cat_out'?
no member named '_convolution' in namespace 'torch'; did you mean 'at::_convolution'?
no member named '_convolution_nogroup' in namespace 'torch'; did you mean 'at::_convolution_nogroup'?
no member named '_copy_from' in namespace 'torch'; did you mean 'at::_copy_from'?
```

Reply:
Ah, I was thinking about this: https://github.com/pytorch/pytorch/blob/812b1ad8695c7bf28a8caa2f001f0ec3c554fbb7/torch/csrc/api/include/torch/types.h#L12-L31

I guess we don't include this anywhere outside of the "C++ API" (for some definition of the C++ API) and I'm not sure if it would cause problems.